nomos-services/blend/src/core/state.rs

```diff
@@ -1,69 +1,382 @@
-}
+    #[derive(Clone, Serialize, Deserialize)]
+    /// Recovery state that is serialized and deserialized to file.
+    ///
+    /// For details about its fields, check [`ServiceState`].
+    pub struct SerializableServiceState<BroadcastSettings> {
+        last_seen_session: u64,
+        spent_core_quota: u64,
+        #[serde(bound(
+            deserialize = "BroadcastSettings: Deserialize<'de> + Eq + core::hash::Hash"
+        ))]
+        unsent_processed_messages: HashSet<ProcessedMessage<BroadcastSettings>>,
+    }
 
-impl ServiceState {
-    pub const fn with_session(session: u64) -> Self {
-        Self {
-            last_seen_session: session,
-            spent_core_quota: 0,
+    impl<BroadcastSettings> SerializableServiceState<BroadcastSettings> {
+        /// Consume the serializable state to create an actual state object, by
+        /// passing it an Overwatch
+        /// [`overwatch::services::state::StateUpdater`].
+        pub fn into_state_with_state_updater<BackendSettings>(
+            self,
+            state_updater: overwatch::services::state::StateUpdater<
+                Option<RecoveryServiceState<BackendSettings, BroadcastSettings>>,
+            >,
+        ) -> ServiceState<BackendSettings, BroadcastSettings> {
+            ServiceState::new(
+                self.last_seen_session,
+                self.spent_core_quota,
+                self.unsent_processed_messages,
+                state_updater,
+            )
         }
     }
 
-    pub const fn last_seen_session(&self) -> u64 {
-        self.last_seen_session
+    impl<BackendSettings, BroadcastSettings> From<ServiceState<BackendSettings, BroadcastSettings>>
+        for SerializableServiceState<BroadcastSettings>
+    {
+        fn from(value: ServiceState<BackendSettings, BroadcastSettings>) -> Self {
+            let (last_seen_session, spent_core_quota, unsent_processed_messages, _) =
+                value.into_components();
+            Self {
+                last_seen_session,
+                spent_core_quota,
+                unsent_processed_messages,
+            }
+        }
     }
+}
 
-    pub const fn spent_core_quota(&self) -> u64 {
-        self.spent_core_quota
+pub use self::service::ServiceState;
+mod service {
+    use core::{
+        fmt::{self, Debug, Formatter},
+        hash::Hash,
+    };
+    use std::collections::HashSet;
+
+    use crate::{
+        core::state::{recovery_state::RecoveryServiceState, state_updater::StateUpdater},
+        message::ProcessedMessage,
+    };
+
+    #[derive(Clone)]
+    /// Recovery state for Blend core service.
+    pub struct ServiceState<BackendSettings, BroadcastSettings> {
+        /// The last session that was saved.
+        last_seen_session: u64,
+        /// The last value for the core quota allowance for the session that is
+        /// tracked.
+        spent_core_quota: u64,
+        unsent_processed_messages: HashSet<ProcessedMessage<BroadcastSettings>>,
```

Comment:
Could we probably replace these three fields with the `SerializableServiceState`?